--- /usr/lib/maliit/plugins/jolla-keyboard.qml	2020-02-11 01:00:00.000000000 +0100
+++ plugin/okboard.qml	2020-04-05 10:37:12.649618020 +0200
@@ -41,6 +41,8 @@
 import org.nemomobile.dbus 2.0
 import org.nemomobile.systemsettings 1.0
 
+import eu.cpbm.okboard 1.0 // okboard
+
 Item {
     id: canvas
 
@@ -66,6 +68,16 @@
 
     onPortraitLayoutChanged: keyboard.updateLayoutIfAllowed(true)
 
+    /* --- okboard begin --- */
+    property bool showPredictionList: layoutRow.layout && !layoutRow.layout.splitActive && (keyboard.curvepreedit || keyboard.curveerror)
+    property InputHandler okbSaveInputHandler
+    property Component okbSaveTopItem
+
+    onShowPredictionListChanged: {
+	keyboard.updateInputHandler();
+    }
+    /* --- okboard end --- */
+
     function updateIMArea() {
         if (!MInputMethodQuick.active)
             return
@@ -308,7 +320,16 @@
                 }
             }
 
-            KeyboardBase {
+	    /* --- okboard begin --- */
+            PredictList {
+                id: curvePredictionList
+            }
+            VerticalPredictList {
+                id: curveVerticalPredictionList
+            }
+	    /* --- okboard end --- */
+
+            CurveKeyboardBase {
                 id: keyboard
 
                 property color popperBackgroundColor: Theme.colorScheme === Theme.LightOnDark
@@ -333,6 +354,7 @@
                                                               : (phone_portrait.visible ? phone_portrait
                                                                                         : phone_landscape.item)
                 layoutChangeAllowed: mode === "common"
+		/* okboard remove
                 thresholdX: swipeGestureIsSafe ? (Theme.startDragDistance * 4.0) : (Theme.startDragDistance * 6.0)
                 thresholdY: Theme.startDragDistance * 1.8
                 swipeEnabled: layoutChangeAllowed && (canvas.layoutModel.enabledCount > 1) && !!layout && layout.allowSwipeGesture
@@ -357,7 +379,8 @@
                     if (gestureInProgress) {
                         layoutRow.updateTransitionDirection(getGestureNextLayoutIndex())
                     }
-                }
+                } */
+
                 onModeChanged: layoutRow.layout.visible = mode === "common"
 
                 function getGestureNextLayoutIndex() {
@@ -466,9 +489,32 @@
                     }
 
                     updateInputHandler()
+		    keyboard.updateCurveContext() // okboard
                 }
 
-                function getInputHandler(layout) {
+		function getInputHandler(layout) {
+		    /* --- okboard begin --- */
+		    // replace InputHandler's topItem with my swipe candidate list
+		    var retInputHandler = getInputHandlerOrig(layout)
+		    if (retInputHandler) {
+			if (retInputHandler.topItem != curvePredictionList) {
+			    // backup original InputHandler topItem because we will need to revert to this
+			    // one when we no longer need to display the swipe candidate list
+			    okbSaveTopItem = retInputHandler.topItem
+			    okbSaveInputHandler = retInputHandler
+			}
+
+			if (showPredictionList) {
+			    retInputHandler.topItem = curvePredictionList
+			} else if (okbSaveTopItem) {
+			    retInputHandler.topItem = okbSaveTopItem  // revert to original topItem
+			}
+		    }
+		    return retInputHandler
+		}
+
+                function getInputHandlerOrig(layout) {
+		    /* --- okboard end --- */
                     var layoutModelItem = layoutModel.get(layout.layoutIndex)
                     if (layoutModelItem === null) {
                         console.warn("could not find layout model item for layout index:", layout.layoutIndex)
@@ -582,8 +628,18 @@
                 }
 
                 Loader {
+		    /* okboard remove
                     sourceComponent: keyboard.inputHandler && layoutRow.layout && layoutRow.layout.splitActive
                                      ? keyboard.inputHandler.verticalItem : null
+		    */
+		    /* --- okboard begin --- */
+		    // this is ridiculously unreadable but does not work from a function
+		    sourceComponent: (layoutRow.layout && layoutRow.layout.splitActive)?(
+			(keyboard.curvepreedit || keyboard.curveerror)?curveVerticalPredictionList:(
+			    keyboard.inputHandler?keyboard.inputHandler.verticalItem:null
+			)
+		    ):null;
+		    /* --- okboard end --- */
                     width: geometry.middleBarWidth
                     height: keyboard.height
                     anchors.horizontalCenter: parent.horizontalCenter
@@ -592,7 +648,7 @@
 
                 KeyboardLayoutSwitchHint {
                     id: switchHint
-                    enabled: keyboard.layout && (keyboard.layout.type !== "hwr") && keyboard.swipeEnabled
+                    enabled: keyboard.layout && (keyboard.layout.type !== "hwr") // okboard remove : && keyboard.swipeEnabled
                     onFinished: closeHintDate.value = Date.now() / 1000 + (2 * 24 * 3600) // Set the close hint date 2 days from now
                     onActiveChanged: if (active) closeHintDate.value = 0 // Clear the other hint when this one is active
                 }
@@ -685,9 +741,11 @@
             ScriptAction {
                 script: {
                     // If there is a stuck transition, stop it
+		    /* okboard remove
                     if (keyboard.gestureInProgress) {
                         keyboard.cancelGesture()
                     }
+		    */
 
                     // Load neighbours of current layouts to ensure smooth swipe
                     _layoutRow.updateLoadersToLayoutAndNeighbours(canvas.activeIndex)
